### DISCUSSION TITLE
How to identify a wildcard from an error message?

### DISCUSSION THREAD

**Discussion by dhosein** [adieyal/sd-dynamic-prompts#834] [Q&A]
I keep running into errors in dyamic prompts that look like this:
Error running process: C:\pinokio\api\stable-diffusion-webui-forge.git\app\extensions\sd-dynamic-prompts\scripts\dynamic_prompting.py0s/it] 
    Traceback (most recent call last):
    
         File "C:\pinokio\api\stable-diffusion-webui-forge.git\app\venv\lib\site-packages\pyparsing\core.py", line 1199, in parse_string
        raise exc.with_traceback(None)
    pyparsing.exceptions.ParseException: Expected end of text, found '_'  (at char 99), (line:1, col:100)
    
    Now... I think I understand that this is somewhere in a wildcard where I have a single _ instead of a double __. 
    
    Question is which? I have about 700 wildcards, some of which have over 100 entries within. I've tried going through all of them manually.. and even creating an excel sheet to try to analyze which wildcard(s) are problematic.. but I'm still getting the error. 
    
    Is there any way from the terminal text to identify which wildcard was being resolved when the error occured? And if not.. shouldn't there be?

**Comment by Hyphona** [comment]
Hello, I made this Python script quickly that will parse all the wildcards for you, create a "start.py" file inside your wildcards directory and paste this code, then open a terminal in the same directory and type "python ./start.py"

Edit:
- Fixed false positive mentionned in this comment https://github.com/adieyal/sd-dynamic-prompts/discussions/834#discussioncomment-11604155

```
import glob
import re

# List all the file
wildcards = glob.glob('./*.txt')

def check_line(line):
    words = line.split()
    for word in words:
        if word.startswith('_') or (word.endswith('_') or word.endswith('_\n')):
            pattern = r'^__.*__(\n)?$'
            if not re.match(pattern, word):
                return False
    return True

# Open each file one by one and check its content line by line
for wildcard in wildcards:
    file = open(wildcard, 'r')
    lines = file.readlines()
    if not lines:
        print(f'The file {0} is empty'.format(file))
    else:
        i = 0
        for line in lines:
            i += 1
            if not check_line(line):
                print(f'Error in {0} at line: {1}'.format(wildcard, i))
    file.close()
```

**Comment by dhosein** [comment]
That's brilliant. I've managed to find a couple of errors using this script. But it does also throw up a number of false positives. 

For example, the following are flagged:
armed with a longsword and a round shield __sword_shield__
swinging a longsword and carrying a round shield __sword_shield__

I did wonder if it's the single _ in the middle of the wildcard name that's doing it, but it also flagged:
hands glowing with healing magic __magic__
armed with a holy quarterstaff __staff__

NB: The bold text is actually written with double-underscore on either side. the display interprets that as bold for some reason.. and I've no idea how to stop it doing that.

**Reply by Hyphona** [reply]
They are flagged because the line end by `__` but doesn't start with `__`, here is a fix:

(It checks each word instead of the whole line)

```
import glob
import re

# List all the file
wildcards = glob.glob('./*.txt')

def check_line(line):
    words = line.split()
    for word in words:
        if word.startswith('_') or (word.endswith('_') or word.endswith('_\n')):
            pattern = r'^__.*__(\n)?$'
            if not re.match(pattern, word):
                return False
    return True

# Open each file one by one and check its content line by line
for wildcard in wildcards:
    file = open(wildcard, 'r')
    lines = file.readlines()
    if not lines:
        print(f'The file {0} is empty'.format(file))
    else:
        i = 0
        for line in lines:
            i += 1
            if not check_line(line):
                print(f'Error in {0} at line: {1}'.format(wildcard, i))
    file.close()
```

**Reply by dhosein** [reply]
I think that's got it. 
Thanks very much.

**Comment by dhosein** [comment]
damn.. I'm still running into this:

pyparsing.exceptions.ParseException: Expected end of text, found '_'  (at char 78), (line:1, col:79)

Really wish the error message would mention which wildcard it was parsing.

**Reply by Hyphona** [reply]
What does your prompt look like?

**Reply by dhosein** [reply]
That's the tricky part. My prompt is simply __picture__
It's a wildcard that breaks down further into __medium__ , __shot__ , __angle__ , __lighting__ , __character__ 
and so on. There's about 700 or so wildcards total (not all used at once of course) and some have 100 or more entries. 

So finding which one(s) have this single '_' is nearly impossible without help. Hence why the script was so useful in eliminating any errors. I found a bunch that I've now resolved and the script finds no more examples of a single '_' at the end or at the beginning of any wildcard reference. But I still get errors when dynamic_prompts is trying to resolve the wildcards. 

The only thing I can think of.. is it complaining about my using single _ to separate words in wildcard names? like __This_is_a_wildcard__ ?
